### PR TITLE
Validation (round 1)

### PR DESCRIPTION
**Pull request by joshmoore**
* [x] add `Tile.validate` and implement `TileSet.validate()` to loop over each tile
 * [ ] propagate `src_doc` to TileSets and Tiles
 * [ ] define reasonable _invalid_ documents

Note: the new test set was generated by spacetx-fov-writer [0.0.2](https://gitlab.com/openmicroscopy/incubator/spacetx-fov-writer/tags/0.0.2) given the input `image&sizeX=8&sizeY=8.fake`.

### CODE REVIEW

**Comment by joshmoore** [comment]
Hmmm.... ok. Needs rethinking since this is currently part of the backend as opposed to object validation:

```
E                   slicedimage.backends._base.ChecksumValidationError: calculated checksum 
(eb69f9cdcf27ac7be3014f7ea209ec7cc289b3af5168ec69e79231900d93d0c7) does not match 
expected checksum (...CORRUPTION...e3014f7ea209ec7cc289b3af5168ec69e79231900d93d0c7)
```

**Comment by joshmoore** [comment]
**TIL:** npyio requires the file handle to have seek. tifffile _inspects_ whether it has seek and uses it if it exists.

**Comment by ttung** [comment]
> TIL: npyio requires the file handle to have seek. tifffile inspects whether it has seek and uses it if it exists.

Not fully correct.  tifffile checks if it has seek, and if it does, it uses it.  If it doesn't, it reads into a BytesIO object. :(

As a result, there is no point in building streaming support.

**Comment by joshmoore** [comment]
> As a result, there is no point in building streaming support.

Doh.

@ttung : is there any form of validation that you can see being of value at the TileSet / Tile level at the moment? Is it worth passing in `src_doc` from `Experiment.from_json` and checking against that?

**Comment by ttung** [comment]
> @ttung : is there any form of validation that you can see being of value at the TileSet / Tile level at the moment? Is it worth passing in src_doc from Experiment.from_json and checking against that?

Yes, I think validation for TileSets and Collections akin to what was built for sptx-format would be great. :)

File: slicedimage/_tile.py [hunk 1]
```diff
@@ -80,3 +125,13 @@ def copy(self, dst_fh):
             self.tile_format = ImageFormat.NUMPY
         else:
             raise RuntimeError("copy can only be called on a tile that hasn't been decoded.")
+
+    def validate(self):
+        """
+        Assert the calculated sha256 matches the one provided on object creation.
+        A missing expected sha256 value is considered non-fatal.
+        """
+        self._load()
```
Comment (ttung) [inline slicedimage/_tile.py]: I think _load will now validate against incorrect checksums.

Comment (joshmoore) [reply slicedimage/_tile.py]: That's definitely what I was seeing in https://github.com/spacetx/slicedimage/pull/61#issuecomment-420242372

**Comment by joshmoore** [comment]
> Yes, I think validation for TileSets and Collections akin to what was built for sptx-format would be great. 

Hmmm... think I could use a quick summary of what you think that entails.

**Comment by joshmoore** [comment]
Closing this in favor of a far future backend. :wink:

**Pull request closed without merging**